src/Agent.Worker/JobRunner.cs

```diff
@@ -250,6 +253,47 @@ public sealed class JobRunner : AgentService, IJobRunner
             }
         }
 
+        private async Task<TaskResult> CompleteJobAsync(IJobServer jobServer, IExecutionContext jobContext, AgentJobRequestMessage message, TaskResult? taskResult = null)
+        {
+            var result = jobContext.Complete(taskResult);
+            var outputVariables = jobContext.Variables.GetOutputVariables();
+            var webApiVariables = outputVariables.ToWebApiVariables();
+            var jobCompletedEvent = new JobCompletedEvent(message.RequestId, message.JobId, result, webApiVariables);
+
+            var completeJobRetryLimit = 5;
+            var exceptions = new List<Exception>();
+            while (completeJobRetryLimit-- > 0)
+            {
+                try
+                {
+                    await jobServer.RaisePlanEventAsync(message.Plan.ScopeIdentifier, message.Plan.PlanType, message.Plan.PlanId, jobCompletedEvent, default(CancellationToken));
+                    return result;
+                }
+                catch (TaskOrchestrationPlanNotFoundException ex)
+                {
+                    Trace.Error($"TaskOrchestrationPlanNotFoundException received, while attempting to raise JobCompletedEvent for job {message.JobId}. Error: {TeamFoundationExceptionFormatter.FormatException(ex, false)}");
+                    return TaskResult.Failed;
+                }
+                catch (TaskOrchestrationPlanSecurityException ex)
+                {
+                    Trace.Error($"TaskOrchestrationPlanSecurityException received, while attempting to raise JobCompletedEvent for job {message.JobId}. Error: {TeamFoundationExceptionFormatter.FormatException(ex, false)}");
```

Comment:
just do ex.tostring(), we never use TeamFoundationExceptionFormatter in this project.